File: 3356. Zero Array Transformation II/Zero Array Transformation II.py
nums = [2,0,2]
queries = [[0,2,1],[0,2,1],[1,1,3]]
def minZeroArray(nums, queries):
    n = len(nums)
    lis = [0]*n
    temp = 0
    index = 0
    while index!=n:
        if (nums[index] - (temp + lis[index])) <= 0:
            temp+=lis[index]
            index+=1
        else:
            break
    else:
        return 0
    for i in range(len(queries)):
        l, r, val = queries[i]
        if r>=index:
            if index>l:
                lis[index] += val
            else:
                lis[l] += val
        if r!=n-1 and r>=index:
            lis[r+1] -= val
        while index!=n:
            if (nums[index] - (temp + lis[index])) <= 0:
                temp+=lis[index]
                index+=1
            else:
                break
        else:
            return i+1
    return -1
print(minZeroArray(nums,queries))

File: 3356. Zero Array Transformation II/ZeroArrayTransformationII.java
public class ZeroArrayTransformationII {
    public int minZeroArray(int[] nums, int[][] queries) {
        int n = nums.length,l,r,v,sum = 0,k = 0;
        int[] arr = new int[n+1];
        for(int i=0;i<n;i++){
            sum+=arr[i];
            while (sum<nums[i]){
                if (k==queries.length){
                    return -1;
                }
                l = queries[k][0];
                r = queries[k][1];
                v = queries[k][2];
                arr[l] += v;
                arr[r+1] -= v;
                if (l<=i && i<=r){
                    sum+=v;
                }
                k+=1;
            }
        }
        return k;
    }
    public static void main(String[] args) {
        ZeroArrayTransformationII minZeroArray = new ZeroArrayTransformationII();
        System.out.println(minZeroArray.minZeroArray(new int[]{2,0,2}, new int[][]{{0,2,1},{0,2,1},{1,1,3}}));
    }
}
